### PR TITLE
Support comments in serde_json.

### PR DESCRIPTION
**Pull request by nathanhammond**
The beginnings of a fully-automated `serde_jsonc` fork?

### CODE REVIEW

File: src/de.rs [hunk 1]
```diff
@@ -251,6 +293,11 @@ impl<'de, R: Read<'de>> Deserializer<R> {
                 Some(b' ') | Some(b'\n') | Some(b'\t') | Some(b'\r') => {
                     self.eat_char();
                 }
+                // Treat comments as whitespace.
+                Some(b'/') => {
+                    self.eat_char();
+                    return self.parse_comment();
+                }
```
Comment (nathanhammond) [inline src/de.rs]: This is the entire integration point into existing code. Given the relatively small integration point I hypothesize that we will be able to maintain this as a floating patch in an automated way, indefinitely.

File: src/de.rs [hunk 2]
```diff
@@ -243,6 +243,48 @@ impl<'de, R: Read<'de>> Deserializer<R> {
...
+
+                // TODO: eat until */.
+                return Ok(Some(b'/'));                
+            }
+
+            // Random bad `/` in a whitespace position.
+            _ => {
+                // Lie, false rewind.
+                // FIXME: Actually rewind?
+                return Ok(Some(b'/'));
```
Comment (chris-olszewski) [inline src/de.rs]: Should we error here instead? `parse_whitespace` is supposed to return the first non-whitespace without consuming it, but this codepath is returning the first non-whitespace *and* consuming it.

File: src/de.rs [hunk 3]
```diff
@@ -243,6 +243,48 @@ impl<'de, R: Read<'de>> Deserializer<R> {
...
+                        }
+                    };
+                }
+            }
+
+            // Block Comment
+            Some(b'*') => {
+                self.eat_char();
+
+                // TODO: eat until */.
```
Comment (chris-olszewski) [inline src/de.rs]: Important TODO

File: src/de.rs [hunk 4]
```diff
@@ -243,6 +243,48 @@ impl<'de, R: Read<'de>> Deserializer<R> {
...
+    fn parse_comment(&mut self) -> Result<Option<u8>> {
+        // Cursor is already after the first slash.
+        match tri!(self.peek()) {
+            // Line Comment
+            Some(b'/') => {
+                self.eat_char();
+
+                // And eat the rest of the line.
+                loop {
+                    match tri!(self.peek()) {
```
Comment (NicholasLYang) [inline src/de.rs:287]: Does this work with a comment followed by EOF instead of a newline?

**Review by NicholasLYang — COMMENTED** [review]
This seems pretty maintainable!

File: src/de.rs (hunk 4 above)
Comment (nathanhammond) [reply src/de.rs:287]: It does! I'll get a test in to demonstrate that.

File: src/de.rs (hunk 2 above)
Comment (nathanhammond) [reply src/de.rs]: Yeah, I know this is cheating. I think the right answer here is to change where we eat the character and peek a second `u8` when necessary.

File: src/de.rs (hunk 3 above)
Comment (nathanhammond) [reply src/de.rs]: YUP! But I'm not 100% sure how to do this. The straightforward version is to look for the `*/` byte sequence at `u8` intervals, but I haven't yet convinced myself that it's impossible to create that same byte sequence at non-character boundaries (e.g. four byte chars containing `00101010 00101111`).

```
*/
\u002A \u002F
00101010 00101111
```

Comment (nathanhammond) [reply src/de.rs]: Okay, I was right, I do need to be charset aware:

```
"\uAA2A\u2F00"
10101010 00101010 00101111 00000000
```

Comment (nathanhammond) [reply src/de.rs]: (This also means that the loop looking for EOL in the line comment is wrong.)

**Pull request closed without merging**